### PR TITLE
[WIP] Add OAuth for Google Photos

### LINKED ISSUE #16: [Flickr] Download by Album should be created as Album in Google Drive
https://developers.google.com/photos/library/guides/manage-albums

### PR DESCRIPTION
**Pull request by thongphp**
- [x] Sync album from input textbox to Google Photo Albums
- [ ] Sync gallery from input textbox to Google Photo Albums
- [x] Create Sync download photos to Google Photo Album
- [x] Remove Flickr download photos to Google Drive

Close #16

### CODE REVIEW

File: app/Services/GooglePhoto.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public const ALBUMS_ENDPOINT = 'https://photoslibrary.googleapis.com/v1/albums';
+
+    /**
+     * Ref: https://developers.google.com/photos/library/reference/rest/v1/albums/create
+     *
+     * @param string $title
+     *
+     * @return object|null
+     */
+    public function createAlbum(string $title): ?object
```
Comment (jooservices) [inline app/Services/GooglePhoto.php:19]: For improvement please make sure title is no longer than 500
`Name of the album displayed to the user in their Google Photos account. This string shouldn't be more than 500 characters.`

File: composer.json [hunk 2]
```diff
@@ -21,6 +21,7 @@
         "guzzlehttp/guzzle": "^6.5",
         "guzzlehttp/oauth-subscriber": "^0.3.0",
         "jenssegers/mongodb": "^3.6",
+        "kamermans/guzzle-oauth2-subscriber": "^1.0",
```
Comment (jooservices) [inline composer.json:24]: Please consider use 1 universal Oauth 2 for "everything". Not specific Google

File: app/Jobs/Flickr/FlickrAlbum.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function handle(): void
+    {
+        $photos = Flickr::get('photosets.getPhotos', ['photoset_id' => $this->album->id]);
+
+        if (!$photos) {
+            return;
+        }
+
+        /** @var FlickrAlbums $repository */
```
Comment (jooservices) [inline app/Jobs/Flickr/FlickrAlbum.php:49]: No need

File: app/Jobs/Flickr/FlickrAlbum.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        if (!$flickrAlbum) {
+            $flickrAlbum = $repository->save((array) $this->album);
+        }
+
+        $googleAlbum = GooglePhoto::createAlbum($this->album->title->_content);
+
+        $flickrAlbum->setAttribute('googleRef', $googleAlbum);
+        $flickrAlbum->save();
+
+//        foreach ($photos->photoset->photo as $photo) {
```
Comment (jooservices) [inline app/Jobs/Flickr/FlickrAlbum.php:62]: After you have done this task please just completely remove these code. ( If we are no longer needed any more )

File: app/Http/Controllers/Auth/GoogleController.php [hunk 5]
```diff
@@ -40,13 +43,15 @@ public function callback()
             return;
         }
 
+        /** @var Oauth $model */
```
Comment (jooservices) [inline app/Http/Controllers/Auth/GoogleController.php:46]: No need

File: app/Jobs/Flickr/FlickrAlbum.php (hunk 4 above)
Comment (thongphp) [reply app/Jobs/Flickr/FlickrAlbum.php:62]: Yep, sure, that's why this PR is still has "Work In Progress" in title.

File: app/Http/Controllers/Auth/GoogleController.php [hunk 6]
```diff
@@ -27,26 +29,35 @@ class GoogleController extends BaseController
...
 
+        $code = $request->get('code');
+
+        /** @var Oauth $model */
         $model = app(Oauth::class);
-        $model->name = 'google';
 
         foreach ($user as $key => $value) {
-            $model->{$key} = $value;
+            $model->setAttribute($key, $value);
```
Comment (jooservices) [inline app/Http/Controllers/Auth/GoogleController.php:55]: Is possible please figure out how to bind array to model ?

File: app/Jobs/Flickr/FlickrAlbum.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        /** @var FlickrAlbums $repository */
+        $repository = app(FlickrAlbums::class);
+        $flickrAlbum = $repository->findByAlbumId($this->album->id);
+
+        if (!$flickrAlbum) {
+            $flickrAlbum = $repository->save((array) $this->album);
+        }
+
+        $googleRef = $flickrAlbum->getAttributeValue('googleRef');
```
Comment (jooservices) [inline app/Jobs/Flickr/FlickrAlbum.php:57]: Please use if (!var = sth) {}

**Pull request merged**